// thorify/dist/utils/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const params_1 = require("./params");
exports.calcIntrinsicGas = function (tx) {
    let totalGas = params_1.params.TxGas;
    // calculate data gas
    if (tx.data) {
        const buffer = new Buffer(exports.sanitizeHex(tx.data), "hex");
        let z = 0;
        let nz = 0;
        for (const byte of buffer) {
            if (byte) {
                nz++;
            }
            else {
                z++;
            }
        }
        totalGas += params_1.params.TxDataZeroGas * z;
        totalGas += params_1.params.TxDataNonZeroGas * nz;
    }
    if (!!tx.to) {
        totalGas += params_1.params.ClauseGas;
    }
    else {
        totalGas += params_1.params.ClauseGasContractCreation;
    }
    return totalGas;
};
exports.toPrefixedHex = function (hexStr) {
    if (hexStr.indexOf("0x") === 0) {
        return hexStr;
    }
    else {
        return "0x" + hexStr;
    }
};
exports.sanitizeHex = function (hexStr) {
    if (hexStr.indexOf("0x") === 0) {
        return hexStr.substr(2);
    }
    else {
        return hexStr;
    }
};
exports.isHex = function (hex) {
    return !!hex && ((typeof hex === "string") && /^(-0x|0x)?[0-9a-f]+$/i.test(hex));
};
exports.newNonce = function () {
    return Math.floor((new Date().getTime() / 0xffff) * Math.random() * 0xffff);
};
exports.toInteger = function (input) {
    const num = Number.parseInt(input);
    if (Number.isInteger(num)) {
        return num;
    }
    else {
        return null;
    }
};
