// thor-model-kit/dist/lib/crypto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
const crypto_1 = require("crypto");
const keccak = require('keccak');
const secp256k1 = require('secp256k1');
const blake = require('blakejs');
/**
 * computes blake2b 256bit hash of given data
 * @param data one or more Buffer | string
 */
function blake2b256(...data) {
    let ctx = blake.blake2bInit(32, null);
    data.forEach(d => {
        if (Buffer.isBuffer(d))
            blake.blake2bUpdate(ctx, d);
        else
            blake.blake2bUpdate(ctx, Buffer.from(d, 'utf8'));
    });
    return new types_1.Bytes32(Buffer.from(blake.blake2bFinal(ctx)));
}
exports.blake2b256 = blake2b256;
/**
 * computes keccak256 hash of given data
 * @param data one or more Buffer | string
 */
function keccak256(...data) {
    let h = keccak('keccak256');
    data.forEach(d => {
        if (Buffer.isBuffer(d))
            h.update(d);
        else
            h.update(Buffer.from(d, 'utf8'));
    });
    return new types_1.Bytes32(h.digest());
}
exports.keccak256 = keccak256;
/** Secp256k1 methods set */
var Secp256k1;
(function (Secp256k1) {
    /** generate private key  */
    function generatePrivateKey() {
        for (;;) {
            let privKey = crypto_1.randomBytes(32);
            if (secp256k1.privateKeyVerify(privKey)) {
                return new types_1.Bytes32(privKey);
            }
        }
    }
    Secp256k1.generatePrivateKey = generatePrivateKey;
    /**
     * derive public key(uncompressed) from private key
     * @param privKey the private key
     */
    function derivePublicKey(privKey) {
        return secp256k1.publicKeyCreate(privKey.bytes, false /* uncompressed */);
    }
    Secp256k1.derivePublicKey = derivePublicKey;
    /**
     * derive Address from public key
     * @param pubKey the public key
     */
    function deriveAddress(pubKey) {
        return new types_1.Address(keccak256(pubKey.slice(1)).bytes);
    }
    Secp256k1.deriveAddress = deriveAddress;
    /**
     * sign a message using elliptic curve algorithm on the curve secp256k1
     * @param msgHash hash of message
     * @param privKey serialized private key
     */
    function sign(msgHash, privKey) {
        let sig = secp256k1.sign(msgHash.bytes, privKey.bytes);
        let packed = Buffer.alloc(65);
        sig.signature.copy(packed);
        packed[64] = sig.recovery;
        return packed;
    }
    Secp256k1.sign = sign;
    /**
     * recovery signature to public key
     * @param msgHash hash of message
     * @param sig signature
     */
    function recover(msgHash, sig) {
        if (sig.length != 65)
            throw new Error('invalid signature');
        let recovery = sig[64];
        if (recovery !== 0 && recovery !== 1)
            throw new Error('invalid signature recovery');
        return secp256k1.recover(msgHash.bytes, sig.slice(0, 64), recovery, false);
    }
    Secp256k1.recover = recover;
})(Secp256k1 = exports.Secp256k1 || (exports.Secp256k1 = {}));
