// thorify/dist/extend/accounts.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3Utils = require("web3-utils");
const debug = require("debug")("thor:injector");
const EthLib = require("eth-lib/lib");
const thor_model_kit_1 = require("thor-model-kit");
const utils = require("../utils");
const extendAccounts = function (web3) {
    // signTransaction supports both callback and promise style
    // tslint:disable-next-line:max-line-length
    web3.eth.accounts.signTransaction = function signTransaction(tx, privateKey, callback) {
        debug("tx to sign: %O", tx);
        const sign = function (tx) {
            return __awaiter(this, void 0, void 0, function* () {
                if (!tx.chainTag) {
                    const chainTag = yield web3.eth.getChainTag();
                    if (chainTag) {
                        tx.chainTag = chainTag;
                    }
                    else {
                        throw new Error("error getting chainTag");
                    }
                }
                if (!tx.blockRef) {
                    const blockRef = yield web3.eth.getBlockRef();
                    if (blockRef) {
                        tx.blockRef = blockRef;
                    }
                    else {
                        throw new Error("error getting blockRef");
                    }
                }
                if (!tx.gas) {
                    const gas = yield web3.eth.estimateGas({
                        from: EthLib.account.fromPrivate(utils.toPrefixedHex(privateKey)).address,
                        to: tx.to ? tx.to : "",
                        value: tx.value ? tx.value : 0,
                        data: utils.isHex(tx.data) ? tx.data : "0x",
                    });
                    if (gas) {
                        tx.gas = gas;
                    }
                    else {
                        throw new Error("error getting gas");
                    }
                }
                if (!tx.nonce) {
                    tx.nonce = utils.newNonce();
                }
                const thorTx = utils.ethToThorTx(tx);
                const priKey = new thor_model_kit_1.Bytes32(Buffer.from(utils.sanitizeHex(privateKey), "hex"));
                thorTx.signature = thor_model_kit_1.Secp256k1.sign(thorTx.signingHash, priKey);
                const rawTx = thorTx.encode();
                const result = {
                    rawTransaction: utils.toPrefixedHex(rawTx.toString("hex")),
                    messageHash: thorTx.signingHash.toString("0x"),
                };
                return result;
            });
        };
        // for supporting both callback and promise
        if (callback instanceof Function) {
            sign(tx).then((ret) => {
                return callback(null, ret);
            }).catch((e) => {
                return callback(e);
            });
        }
        else {
            return sign(tx);
        }
    };
    web3.eth.accounts.recoverTransaction = function recoverTransaction(encodedRawTx) {
        const values = EthLib.RLP.decode(encodedRawTx);
        const signingDataHex = EthLib.RLP.encode(values.slice(0, 9));
        const signingHash = utils.hash(Buffer.from(utils.sanitizeHex(signingDataHex), "hex"));
        const signature = values[9];
        const singingHashBuffer = Buffer.from(utils.sanitizeHex(signingHash), "hex");
        const signatureBuffer = Buffer.from(utils.sanitizeHex(signature), "hex");
        const address = utils.recover(singingHashBuffer, signatureBuffer);
        return address;
    };
    web3.eth.accounts.hashMessage = function hashMessage(data) {
        const message = web3Utils.isHexStrict(data) ? web3Utils.hexToBytes(data) : data;
        const messageBuffer = Buffer.from(message);
        return utils.hash(messageBuffer);
    };
    web3.eth.accounts.sign = function sign(data, privateKey) {
        const hash = this.hashMessage(data);
        const hashBuffer = Buffer.from(utils.sanitizeHex(hash), "hex");
        const privateKeyBuffer = Buffer.from(utils.sanitizeHex(privateKey), "hex");
        const signature = utils.sign(hashBuffer, privateKeyBuffer);
        return {
            message: data,
            messageHash: utils.toPrefixedHex(hash),
            signature: utils.toPrefixedHex(signature),
        };
    };
    web3.eth.accounts.recover = function recover(message, signature, preFixed) {
        const args = [].slice.apply(arguments);
        if (web3Utils._.isObject(message)) {
            return this.recover(message.messageHash, message.signature, true);
        }
        if (!preFixed) {
            message = this.hashMessage(message);
        }
        const hexBuffer = Buffer.from(utils.sanitizeHex(message), "hex");
        const signatureBuffer = Buffer.from(utils.sanitizeHex(signature), "hex");
        return utils.recover(hexBuffer, signatureBuffer);
    };
};
exports.extendAccounts = extendAccounts;
